File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isValid(root, low, high) -> bool:
            # Base case. None is True --> every leaf node is a valid BST tree
            if root is None: return True
            
            if root.val >= high or root.val <= low:
                return False
        
            return isValid(root.left, low, root.val) and isValid(root.right, root.val, high)
        
        return isValid(root, -float(inf), float(inf))

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    double pre = -INFINITY;
    bool isValidBST(TreeNode* root) {
        if (!root) return true;
        
        // in-order traverse BST
        bool l = isValidBST(root->left);
        if(root->val <= pre) {
            return false;
        }
        pre = root->val;
        bool r = isValidBST(root->right);
        return l && r;
    }
};